### PR TITLE
Stop running prebuilds for inactive projects (10+ weeks)

### PR DESCRIPTION
**Pull request by jankeromnes**
## Description

Stop running prebuilds for projects that did not start a workspace in the last 10+ weeks.

## Related Issue(s)

Fixes https://github.com/gitpod-io/gitpod/issues/8911
Drive-by fixes prebuild rate limits again

## How to test

1. Add a Project
2. Verify that Prebuilds are triggered successfully and workspaces can be started
3. Verify that `lastWorkspaceStart` and `lastWebhookReceived` are properly set in `d_b_project_usage`
4. Manually set `lastWorkspaceStart` to a date more than 10 weeks ago in `d_b_project_usage`
5. Trigger new Prebuilds -- verify that they now get auto-cancelled on start due to Project inactivity

## Release Notes

```release-note
Stop running prebuilds for projects that did not start a workspace in the last 10+ weeks
```

### CODE REVIEW

**Comment by gtsiolis** [comment]
**fyi**: I've opened a follow up issue with a feature request to surface this automatic pause action better on the dashboard as this could be confusing to users coming back to inactive projects in the dashboard, see https://github.com/gitpod-io/gitpod/issues/9232. Feedback is welcome!

Cc @jankeromnes @easyCZ @jldec

**Comment by jankeromnes** [comment]
Looks like this now when a project becomes inactive, but still triggers new prebuilds:

<table><tr><td>
Screenshot 2022-04-12 at 11 30 54
</td><td>
Screenshot 2022-04-12 at 11 31 06
</td></tr></table>

File: components/gitpod-db/src/typeorm/entity/db-project-usage.ts
Comment (easyCZ) [inline components/gitpod-db/src/typeorm/entity/db-project-usage.ts:17]: What's historically the reason for using varchar to store timestamps? Is it possible to use a DATETIME?

Comment (jankeromnes) [reply components/gitpod-db/src/typeorm/entity/db-project-usage.ts:17]: Good question -- I don't have a strong opinion here, and was just following established practice 😊

Comment (jankeromnes) [reply components/gitpod-db/src/typeorm/entity/db-project-usage.ts:17]: > The reason I ask is that when prototyping an ORM for Golang, Varchar timestamps make it much more complicated - this is further coupled with the somewhat non-standard usage of ISO8601 in javascript to represent a timestamp, which isn't quite RFC3339 compatible.

I think your concerns are valid, and there is definitely room for improvement in our DB types. 👍

> Definitely not trying to block this PR on this, mostly looking to understand historical reasoning for the varchar choice. Feel free to ignore this comment to land the change.

Thanks! I would love to, but first I need a review. 😇

Comment (easyCZ) [reply components/gitpod-db/src/typeorm/entity/db-project-usage.ts:17]: Still reviewing :)

Comment (jankeromnes) [reply components/gitpod-db/src/typeorm/entity/db-project-usage.ts:17]: That's awesome, thanks a bunch 🙏

I'll try to practice our new review process by assigning this Pull Request to you then (but please feel free to un-assign yourself again if that didn't make sense!)

**Comment by easyCZ** [comment]
One alternative approach we could consider is to load the `d_b_project_usage` async from existing tables (Prebuild, WorkspaceInstance) to treat it like more of a cache. Doing this would achieve the following:
1. The handling code for a project doesn't need to be aware of the usage tracking - this eliminates opportunity for a bug where we forget to add the call to track
2. It derives the usage from a single place, rather than multiple
3. It makes the update process async which does not impact our latency on project related calls (the setting of the value).

WDYT?

File: components/gitpod-db/src/typeorm/entity/db-project-usage.ts
Comment (easyCZ) [inline components/gitpod-db/src/typeorm/entity/db-project-usage.ts:18]: Is this field strictly needed for our first version? We do set it, but we're not using it in read paths.

File: components/gitpod-db/src/typeorm/migration/1649667202321-ProjectUsage.ts
Comment (easyCZ) [inline components/gitpod-db/src/typeorm/migration/1649667202321-ProjectUsage.ts:12]: Here's we're setting a limit on `projectId` of 36 chars, should this also be reflected in the `precision: 36` field on the Column of the ORM model?

File: components/server/ee/src/prebuilds/github-enterprise-app.ts
Comment (easyCZ) [inline components/server/ee/src/prebuilds/github-enterprise-app.ts:121]: At the point of this call, how do we know that the project specified in the payload in fact belongs to the correct user? Is it possible to poison our data across projects here?

Mostly lacking context, so trying to understand better how the payload from the webhook is authhorized

**Review by easyCZ — APPROVED** [review]
PR looks good, and worked for me based on the test specified. Leaving a couple of questions for my own context. Approving with a hold.

/hold

Posting our slack conversation about splitting PRs here as well for visibility.
For the PR, with regards to tiny PRs. I'd split it into the following:

1. Add a method to call to updateProjectUsage but don't write to the DB, only increment a metric
2. Create the table, it's unused
Get data that this works
3. Wire them together - updateProjectUsage actually writes to the DB

File: components/gitpod-db/src/typeorm/entity/db-project-usage.ts
Comment (jankeromnes) [reply components/gitpod-db/src/typeorm/entity/db-project-usage.ts:18]: Good catch, it's not strictly needed here indeed. However, since I modified very related code, I thought this small drive-by change could be a good first step toward fixing https://github.com/gitpod-io/gitpod/issues/7010

Comment (easyCZ) [reply components/gitpod-db/src/typeorm/entity/db-project-usage.ts:18]: I'll leave it up to you. 

For surfacing which webhook we've received, I'd expose it as a first class-table with webhooks (and webhook payloads), such that we can show what we've received in the webhook as well (would definitely be useful for our debugging).

Comment (jankeromnes) [reply components/gitpod-db/src/typeorm/entity/db-project-usage.ts:18]: Good idea. I'll leave this as is, and we can decide later whether we want a full-fledged table of webhook events, or whether showing a simple timestamp of the last event is good enough. 🛹

File: components/gitpod-db/src/typeorm/migration/1649667202321-ProjectUsage.ts
Comment (jankeromnes) [reply components/gitpod-db/src/typeorm/migration/1649667202321-ProjectUsage.ts:12]: I'm not too familiar with the intricacies of our ORM model, so I wouldn't know the pros and cons of doing this.

Briefly looking through our code, it seems that we typically use `precision` for timestamps, but not for IDs. If you think this should be changed, could you please open an issue for this? 🙏

File: components/server/ee/src/prebuilds/github-enterprise-app.ts
Comment (jankeromnes) [reply components/server/ee/src/prebuilds/github-enterprise-app.ts:121]: > At the point of this call, how do we know that the project specified in the payload in fact belongs to the correct user?

That's a good question. I don't know about webhook events from the GitHub App, but for GitLab, we look at the provided (user + project) token in order to make sure that the project specified in the payload actually belongs to the user who installed this webhook:

https://github.com/gitpod-io/gitpod/blob/e8fa6d20d2c18cdb0f02ff2cfd483388c5c5dac9/components/server/ee/src/prebuilds/gitlab-app.ts#L61-L97

We do something similar for GitHub Enterprise webhooks, where the secret token is not sent as part of the webhook payload, but is used to sign the webhook payload (which we can verify against the user's tokens):

https://github.com/gitpod-io/gitpod/blob/e8fa6d20d2c18cdb0f02ff2cfd483388c5c5dac9/components/server/ee/src/prebuilds/github-enterprise-app.ts#L89-L106

> Is it possible to poison our data across projects here?

I don't think it is possible to send forged webhook events to Gitpod.

However, your question does remind me of an issue where Pull Requests can come from forks, but trigger a Prebuild for the main repository, which can be problematic in some cases: https://github.com/gitpod-io/security/issues/26

**Comment by jankeromnes** [comment]
Many thanks for your very helpful review and suggestions @easyCZ! 🙏

> One alternative approach we could consider is to load the `d_b_project_usage` async from existing tables (Prebuild, WorkspaceInstance) to treat it like more of a cache. Doing this would achieve the following:
> 1. The handling code for a project doesn't need to be aware of the usage tracking - this eliminates opportunity for a bug where we forget to add the call to track
> 2. It derives the usage from a single place, rather than multiple
> 3. It makes the update process async which does not impact our latency on project related calls (the setting of the value).
> 
> WDYT?

I like the idea of doing these things more async, because it could make our system simpler and much less subtle/magical.

(For example, I initially updated `lastWorkspaceStart` only when starting from a prebuild, so starting a workspace not from a prebuild wouldn't count -- I've since added the missing update code, but it's pretty easy to imagine a future refactor where someone accidentally forgets one of these updates again.)

However, it feels like we're lacking some good framework code for such asynchronous jobs that regularly process large amounts of data in the background. I agree that we have to start somewhere, but I'm a bit hesitant to volunteer this Pull Request to be the first "async background processing job", given that it can already provide value to users as is. 🛹

---

I've also replied to your questions in-line. Many thanks for diving deeper into all these systems! I find your questions and instincts for improvements quite relevant, and am keen to help improve things or provide context whever I can. 🙂

Also, removing the hold, as I think this is good to go as is. Happy to improve anything in follow-ups. 🚀

/unhold

**Pull request merged**
